pennylane/tape/qscript.py

```diff
@@ -395,41 +378,57 @@ def _update(self):
         """Update all internal metadata regarding processed operations and observables"""
         self._graph = None
         self._specs = None
-        self._update_circuit_info()  # Updates wires, num_wires; O(ops+obs)
-        self._update_par_info()  # Updates _par_info; O(ops+obs)
-
-        self._update_observables()  # Updates _obs_sharing_wires and _obs_sharing_wires_id
 
-    def _update_circuit_info(self):
-        """Update circuit metadata
+        try:
+            # Invalidate cached properties so they get recalculated
+            del self.wires
+            del self.par_info
+        except AttributeError:
+            pass
 
-        Sets:
-            wires (~.Wires): Wires
-            num_wires (int): Number of wires
-        """
-        self.wires = Wires.all_wires(dict.fromkeys(op.wires for op in self))
-        self.num_wires = len(self.wires)
+    @cached_property
+    def wires(self) -> Wires:
+        "Wires used in the quantum script process"
+        return Wires.all_wires(dict.fromkeys(op.wires for op in self))
 
-    def _update_par_info(self):
-        """Update the parameter information list. Each entry in the list with an operation and an index
-        into that operation's data.
+    @property
+    def num_wires(self) -> int:
+        """Number of wires in the quantum script process"""
+        return len(self.wires)
 
-        Sets:
-            _par_info (list): Parameter information
+    @cached_property
+    def par_info(self):
+        """list[dict[str, Operator or int]]: Parameter information.
+        Values are dictionaries containing the corresponding operation and operation parameter index.
```

Comment:
Potentially worth sticking a example here for reference.

Reply:
True. I was also confused when I first saw the name with what it actually does. Just added the example 😄